PR for SRVKE-1692: Add authorization and eventpolicy subsection in the Eventing section (#97655)
openshift/openshift-docs: modules/serverless-event-auth-applying-eventpolicy.adoc

kind: Broker
metadata:
  name: broker
  namespace: namespace-1
---
# PingSource in namespace-1
apiVersion: sources.knative.dev/v1
kind: PingSource
metadata:
  name: pingsource-1
  namespace: namespace-1
spec:
  data: '{"message": "Hi from pingsource-1 from namespace-1"}'
  schedule: '*/1 * * * *'
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1
      kind: Broker
      name: broker
      namespace: namespace-1
---
# PingSource in namespace-2
apiVersion: sources.knative.dev/v1
kind: PingSource
metadata:
  name: pingsource-2
  namespace: namespace-2
spec:
  data: '{"message": "Hi from pingsource-2 from namespace-2"}'
  schedule: '*/1 * * * *'
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1
      kind: Broker
      name: broker
      namespace: namespace-1
----

. Create an event-display service to show incoming events and add a Trigger to connect it to the Broker as per shown in the following example:
+
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: event-display
  namespace: namespace-1
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/min-scale: "1"
    spec:
      containers:

>       - image: gcr.io/knative-releases/knative.dev/eventing/cmd/event_display

Reviewer: maybe we can use a plain k8s service and deployment to not require serving at this point

Author: I have added a plain Kubernetes Deployment + Service version (no Knative Serving), with the Trigger still subscribing to the Service. Please check and let me know

---
apiVersion: eventing.knative.dev/v1
kind: Trigger
metadata:
  name: trigger
  namespace: namespace-1
spec:
  broker: broker
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: event-display
----
+
At this stage, because OIDC is disabled and no EventPolicy is applied, the event-display service shows events from both PingSources.

. Enable OIDC in Knative Eventing and create an EventPolicy by executing the following example command:
+
[source,terminal]
----
$ $ oc -n knative-eventing patch cm config-features \
  --type merge \
  -p '{"data":{"authentication-oidc":"enabled"}}'
----

> . Once OIDC is enabled, create an EventPolicy that authorizes only the Pin`gSource from `namespace-2` as per shown in the following example:

Reviewer: ```suggestion
. Once OIDC is enabled, create an EventPolicy that authorizes only the PingSource from `namespace-2` as per shown in the following example:
```

---
# PingSource in namespace-1
apiVersion: sources.knative.dev/v1
kind: PingSource
metadata:
  name: pingsource-1
  namespace: namespace-1
spec:
  data: '{"message": "Hi from pingsource-1 from namespace-1"}'
  schedule: '*/1 * * * *'
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1
      kind: Broker
      name: broker
      namespace: namespace-1
---
# PingSource in namespace-2
apiVersion: sources.knative.dev/v1
kind: PingSource
metadata:
  name: pingsource-2
  namespace: namespace-2
spec:
  data: '{"message": "Hi from pingsource-2 from namespace-2"}'
  schedule: '*/1 * * * *'
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1
      kind: Broker
      name: broker
      namespace: namespace-1
----

. Create an event-display service to show incoming events and add a Trigger to connect it to the Broker as per shown in the following example:
+
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: event-display
  namespace: namespace-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: event-display
  template:
    metadata:
      labels:
        app: event-display
    spec:
      containers:
        - name: event-display

>           image: gcr.io/knative-releases/knative.dev/eventing/cmd/event_display

Reviewer: ```suggestion
          image: gcr.io/knative-releases/knative.dev/eventing/cmd/event_display
          ports:
          - containerPort: 8080
```

---
apiVersion: v1
kind: Service
metadata:
  name: event-display
  namespace: namespace-1
spec:
  selector:
    app: event-display
  ports:

>     - name: http

Reviewer: ```suggestion
    - name: http
      port: 80
      targetPort: 8080
```

we need a port and target port (as event_display is running on 8080) on the service

Author: Added

. Enable OIDC in Knative Eventing and create an EventPolicy by executing the following example command:
+
[source,terminal]
----
$ $ oc -n knative-eventing patch cm config-features \
  --type merge \
  -p '{"data":{"authentication-oidc":"enabled"}}'
----

. Once OIDC is enabled, create an EventPolicy that authorizes only the PingSource from `namespace-2` as per shown in the following example:
+
[source,yaml]
----
apiVersion: eventing.knative.dev/v1alpha1
kind: EventPolicy
metadata:
  name: event-policy
  namespace: namespace-1
spec:
  to:
    - ref:
        apiVersion: eventing.knative.dev/v1
        kind: Broker
        name: broker
  from:
    - ref:
        apiVersion: sources.knative.dev/v1
        kind: PingSource
        name: pingsource-2
        namespace: namespace-2
----

.Verification

. Verify the EventPolicy to check the Broker status by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 get broker broker -o yaml
----

. View logs from the event-display service to confirm only `pingsource-2` events arrive by executing the following command:
+
[source,terminal]
----

> $ oc -n namespace-1 logs -f -l serving.knative.dev/service=event-display

Reviewer: ```suggestion
$ oc -n namespace-1 logs -f -l app=event-display
```
those labels don't work anymore, as we switch to the raw deployment.

----

. Delete the EventPolicy by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 delete eventpolicy event-policy
----

. Check the Broker status to confirm it has returned to the default `allow-same-namespace` mode by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 get broker broker -o yaml
----

. View the event-display service logs to confirm that only `pingsource-1` events from the same namespace appear by executing the following command:
+
[source,terminal]
----

> $ oc -n namespace-1 logs -f -l serving.knative.dev/service=event-display

Reviewer: ```suggestion
$ oc -n namespace-1 logs -f -l app=event-display
```
same as above

+
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: event-display
  namespace: namespace-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: event-display
  template:
    metadata:
      labels:
        app: event-display
    spec:
      containers:
        - name: event-display
          image: gcr.io/knative-releases/knative.dev/eventing/cmd/event_display
          ports:
          - containerPort: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: event-display
  namespace: namespace-1
spec:
  selector:
    app: event-display
  ports:
    - name: http
    port: 80
      targetPort: 8080
---
apiVersion: eventing.knative.dev/v1
kind: Trigger
metadata:
  name: trigger
  namespace: namespace-1
spec:
  broker: broker
  subscriber:
    ref:
      apiVersion: v1
      kind: Service
      name: event-display
----
+
At this stage, because OIDC is disabled and no EventPolicy is applied, the event-display service shows events from both PingSources.

. Enable OIDC in Knative Eventing and create an EventPolicy by executing the following example command:
+
[source,terminal]
----

> $ $ oc -n knative-eventing patch cm config-features \

Reviewer 2: ```suggestion
$ oc -n knative-eventing patch cm config-features \
```

(one extra dollar)

  --type merge \

>   -p '{"data":{"authentication-oidc":"enabled"}}'

Author: Done

  namespace: namespace-1
spec:
  to:
    - ref:
        apiVersion: eventing.knative.dev/v1
        kind: Broker
        name: broker
  from:
    - ref:
        apiVersion: sources.knative.dev/v1
        kind: PingSource
        name: pingsource-2
        namespace: namespace-2
----

.Verification

. Verify the EventPolicy to check the Broker status by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 get broker broker -o yaml
----

. View logs from the event-display service to confirm only `pingsource-2` events arrive by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 logs -f -l app=event-display
----

. Delete the EventPolicy by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 delete eventpolicy event-policy
----

. Check the Broker status to confirm it has returned to the default `allow-same-namespace` mode by executing the following command:
+
[source,terminal]
----
$ oc -n namespace-1 get broker broker -o yaml
----

. View the event-display service logs to confirm that only `pingsource-1` events from the same namespace appear by executing the following command:
+
[source,terminal]
----

> $ oc -n namespace-1 logs -f -l serving.knative.dev/service=event-display

Reviewer 2: ```suggestion
$ oc -n namespace-1 logs -f -l app=event-display
```

creydr's original suggestion was marked as resolved, but not applied here
